### ISSUE TITLE
Issue with output from snippy

### ISSUE THREAD

**Issue by bradle01** [bactopia/bactopia#309]
**Description**
I am running bactopia2.1.0 on a cluster, and everything is working except the variant call, it seems to try and call the variants a few times before giving up and generating an error

**Steps to Reproduce**
```
 bactopia --R1 UMWA3271_R1.fastq.gz --R2 UMWA3271_R2.fastq.gz --datasets /project/umw_evan_bradley/bin/datasets --species "Klebsiella pneumoniae" --outdir /project/bactopia-outputs

!!ERROR!!
```
Caused by:
  Process `BACTOPIA:CALL_VARIANTS (UMWA3271 - refseq-genomes)` terminated with an error exit status (1)

Command executed:

  REFERENCE="refseq-genomes.msh"
  REFERENCE_NAME="refseq-genomes"
  if [[ "refseq-genomes.msh" == "refseq-genomes.msh" ]]; then
      # Get Mash distance (For paired-end, use just R1)
      mash dist -k 31 -s 10000 -t UMWA3271_R1.fastq.gz ${REFERENCE} | grep -v "query" | sort -k 2,2 > distances.txt
  
      # Pick top genome and download
      printf "accession\tdistance\tlatest_accession\tupdated\n" > mash-dist.txt
      select-references.py distances.txt 1  >> mash-dist.txt
      grep -v distance mash-dist.txt | cut -f3 > download-list.txt
      ncbi-genome-download bacteria -l complete -o ./ -F genbank -p 1 -A download-list.txt             -r 3 
  
      # Move and uncompress genomes
      mkdir genbank_temp
      find refseq -name "*.gbff.gz" | xargs -I {} mv {} genbank_temp/
      rename 's/(GC[AF]_\d+).*/$1/' genbank_temp/*
      ls genbank_temp/ | xargs -I {} sh -c 'gzip -cd genbank_temp/{} > {}.gbk'
      rm -rf genbank_temp/ refseq/
  
      # Update variables
      REFERENCE=$(ls *.gbk)
      REFERENCE_NAME=$(basename ${REFERENCE} .gbk)
  
      # Capture version
      mash --version > mash.version.txt 2>&1
      ncbi-genome-download --version > ncbi-genome-download.version.txt 2>&1
  fi
  
  snippy --R1 UMWA3271_R1.fastq.gz --R2 UMWA3271_R2.fastq.gz         --ref ${REFERENCE}         --cpus 1         --ram 15         --outdir ${REFERENCE_NAME}         --prefix UMWA3271         --mapqual 60         --basequal 13         --mincov 10         --minfrac 0         --minqual 100         --maxsoft 10  
  mv ${REFERENCE_NAME}/UMWA3271.log ./
  
  # Add GenBank annotations
  vcf-annotator ${REFERENCE_NAME}/UMWA3271.vcf ${REFERENCE} > ${REFERENCE_NAME}/UMWA3271.annotated.vcf
  
  # Get per-base coverage
  grep "^##contig" ${REFERENCE_NAME}/UMWA3271.vcf > ${REFERENCE_NAME}/UMWA3271.full-coverage.txt
  genomeCoverageBed -ibam ${REFERENCE_NAME}/UMWA3271.bam -d >> ${REFERENCE_NAME}/UMWA3271.full-coverage.txt
  cleanup-coverage.py ${REFERENCE_NAME}/UMWA3271.full-coverage.txt > ${REFERENCE_NAME}/UMWA3271.coverage.txt
  rm ${REFERENCE_NAME}/UMWA3271.full-coverage.txt
  
  # Mask low coverage regions
  mask-consensus.py UMWA3271 ${REFERENCE_NAME}         ${REFERENCE_NAME}/UMWA3271.consensus.subs.fa         ${REFERENCE_NAME}/UMWA3271.subs.vcf         ${REFERENCE_NAME}/UMWA3271.coverage.txt         --mincov 10 > ${REFERENCE_NAME}/UMWA3271.consensus.subs.masked.fa
  
  # Clean Up
  rm -rf ${REFERENCE_NAME}/reference ${REFERENCE_NAME}/ref.fa* ${REFERENCE_NAME}/UMWA3271.vcf.gz*
  
  if [[ "refseq-genomes.msh" == "refseq-genomes.msh" ]]; then
      mv distances.txt ${REFERENCE_NAME}/mash-distances.txt
      mv mash-dist.txt ${REFERENCE_NAME}/mash-selected.txt
      mv download-list.txt ${REFERENCE_NAME}/mash-downloaded.txt
      mv ${REFERENCE} ${REFERENCE_NAME}/
  fi
  
  if [[ false == "false" ]]; then
      find ${REFERENCE_NAME}/ -type f |             grep -v -E "\.bam$|\.bai$|\.log$|\.txt$|\.html$|\.tab$" |             xargs -I {} pigz -n --best -p 1 {}
      pigz -n --best -p 1 ${REFERENCE_NAME}/UMWA3271.coverage.txt
  fi
  
  mkdir results
  mv ${REFERENCE_NAME}/ results/
  
  # Capture versions
  cat <<-END_VERSIONS > versions.yml
  "BACTOPIA:CALL_VARIANTS":
      bedtools: $(echo $(bedtools --version 2>&1) | sed 's/bedtools v//')
      mash: $(echo $(mash --version 2>&1))
      ncbi-genome-download: $(echo $(ncbi-genome-download --version 2>&1))
      pigz: $(echo $(pigz --version 2>&1) | sed 's/pigz //')
      snippy: $(echo $(snippy --version 2>&1) | sed 's/snippy //')
      vcf-annotator: $(echo $(vcf-annotator --version 2>&1) | sed 's/vcf-annotator.py //')
  END_VERSIONS

Command exit status:
  1

Command output:
  (empty)

Command error:
         total no. variants observed              : 408
         total no. reference observed             : 0
  
  Time elapsed: 0.00s
  
  [18:40:48] Running: snpEff ann -noLog -noStats -no-downstream -no-upstream -no-utr -c reference/snpeff.config -dataDir . ref UMWA3271.filt.vcf > UMWA3271.vcf 2>> UMWA3271.log
  [18:40:51] Running: /project/umw_evan_bradley/bin/conda/envs/bactopia2.1.0/share/bactopia-2.1.x/conda/envs/call_variants-2.1.x/bin/snippy-vcf_to_tab --gff reference/ref.gff --ref reference/ref.fa --vcf UMWA3271.vcf > UMWA3271.tab 2>> UMWA3271.log
  [18:40:54] Running: /project/umw_evan_bradley/bin/conda/envs/bactopia2.1.0/share/bactopia-2.1.x/conda/envs/call_variants-2.1.x/bin/snippy-vcf_extract_subs UMWA3271.filt.vcf > UMWA3271.subs.vcf 2>> UMWA3271.log
  [18:40:54] Running: bcftools convert -Oz -o UMWA3271.vcf.gz UMWA3271.vcf 2>> UMWA3271.log
  [18:40:54] Running: bcftools index -f UMWA3271.vcf.gz 2>> UMWA3271.log
  [18:40:54] Running: bcftools consensus --sample GCF_022759625 -f reference/ref.fa -o UMWA3271.consensus.fa UMWA3271.vcf.gz 2>> UMWA3271.log
  [18:40:55] Running: bcftools convert -Oz -o UMWA3271.subs.vcf.gz UMWA3271.subs.vcf 2>> UMWA3271.log
  [18:40:55] Running: bcftools index -f UMWA3271.subs.vcf.gz 2>> UMWA3271.log
  [18:40:55] Running: bcftools consensus --sample GCF_022759625 -f reference/ref.fa -o UMWA3271.consensus.subs.fa UMWA3271.subs.vcf.gz 2>> UMWA3271.log
  [18:40:55] Running: rm -f UMWA3271.subs.vcf.gz UMWA3271.subs.vcf.gz.csi UMWA3271.subs.vcf.gz.tbi 2>> UMWA3271.log
  [18:40:55] Generating reference aligned/masked FASTA relative to reference: UMWA3271.aligned.fa
  [18:41:02] Marked 5403 heterozygous sites with 'n'
  [18:41:03] Creating extra output files: BED GFF CSV TXT HTML
  [18:41:03] Identified 408 variants.
  [18:41:03] Result folder: GCF_022759625
  [18:41:03] Result files:
  [18:41:03] * GCF_022759625/UMWA3271.aligned.fa
  [18:41:03] * GCF_022759625/UMWA3271.bam
  [18:41:03] * GCF_022759625/UMWA3271.bam.bai
  [18:41:03] * GCF_022759625/UMWA3271.bed
  [18:41:03] * GCF_022759625/UMWA3271.consensus.fa
  [18:41:03] * GCF_022759625/UMWA3271.consensus.subs.fa
  [18:41:03] * GCF_022759625/UMWA3271.csv
  [18:41:03] * GCF_022759625/UMWA3271.filt.vcf
  [18:41:03] * GCF_022759625/UMWA3271.gff
  [18:41:03] * GCF_022759625/UMWA3271.html
  [18:41:03] * GCF_022759625/UMWA3271.log
  [18:41:03] * GCF_022759625/UMWA3271.raw.vcf
  [18:41:03] * GCF_022759625/UMWA3271.subs.vcf
  [18:41:03] * GCF_022759625/UMWA3271.tab
  [18:41:03] * GCF_022759625/UMWA3271.txt
  [18:41:03] * GCF_022759625/UMWA3271.vcf
  [18:41:03] * GCF_022759625/UMWA3271.vcf.gz
  [18:41:03] * GCF_022759625/UMWA3271.vcf.gz.csi
  [18:41:03] Walltime used: 5 minutes, 38 seconds
  [18:41:03] Have a suggestion? Tell me at https://github.com/tseemann/snippy/issues
  [18:41:03] Done.
  Traceback (most recent call last):
    File "/project/umw_evan_bradley/bin/conda/envs/bactopia2.1.0/share/bactopia-2.1.x/conda/envs/call_variants-2.1.x/bin/vcf-annotator", line 398, in <module>
      annotator.write_vcf(args.output)
    File "/project/umw_evan_bradley/bin/conda/envs/bactopia2.1.0/share/bactopia-2.1.x/conda/envs/call_variants-2.1.x/bin/vcf-annotator", line 192, in write_vcf
      self.__vcf.write_vcf(output)
    File "/project/umw_evan_bradley/bin/conda/envs/bactopia2.1.0/share/bactopia-2.1.x/conda/envs/call_variants-2.1.x/bin/vcf-annotator", line 357, in write_vcf
      vcf_writer = vcf.Writer(open(output, 'w'), self.reader)
  NotADirectoryError: [Errno 20] Not a directory: '/dev/stdout'

**Execution Environment**

- Bactopia Version: bactopia 2.1.0
- OS: RedHat7
- Environment: conda environment on a cluster

**Additional Information**

It looks like it found the appropriate reference genome and seems to have generated all the of the files from snippy, except it then appears to be looking for something in the root directory, which I can't access because this is a shared cluster, can it be redirected to a local directory?

**Comment by rpetit3** [comment]
Hi @bradle01,

Looks like the issue is here `NotADirectoryError: [Errno 20] Not a directory: '/dev/stdout'`

I can modify things so that it doesn't use `/dev/stdout` which I think should be fine. Let me know push a change to the dev branch. 

Would you be interested in testing?

Cheers,
Robert

**Comment by rpetit3** [comment]
Ok! I think I fixed this. If you interested you can give my dev build a try:

```
mamba install -n bactopia-dev -c rpetit3 -c conda-forge -c bioconda bactopia
```

**Comment by bradle01** [comment]
Your update seems to have solved the issue, I ran bactopia-dev on the same genomes and it produced the snippy output, thanks!

**Comment by rpetit3** [comment]
Hi @bradle01 

I released v2.1.0 (https://github.com/bactopia/bactopia/releases/tag/v2.1.0) which should fix includes the changes from the dev build you were using. If your problem persists, please feel free to reopen this issue!

If you have some time to spare, please consider filling out my Bactopia survey. I use the feedback to direct the future developments of Bactopia. (If you've already filled it out, thank you so much!)

https://forms.gle/PTgUxEBwCsCdbQjr7

Cheers,
Robert
